Create manager for DigitalIO

danielb987 opened the pull request:
I want managers for DigitalIO, AnalogIO and StringIO.

This PR is a try on creating a DigitalIOManager that also is able to get turnouts, sensors and lights.

For the GeneralManagerTest, I needed to create a connection and wasn't sure on how to do it, so maybe there are better was to do it?

Any thoughts of this before I continue with AnalogIO and StringIO?

rhwood commented:
Given the complexity of what you are doing, do not rely on DefaultInstanceInitializer to create your default instance, but create an instance of InstanceInitializer specific to the class for that purpose.

pabender commented:
Sent from my iPad

> On Mar 31, 2019, at 12:21 PM, Daniel <notifications@github.com> wrote:
> 
> I want managers for DigitalIO, AnalogIO and StringIO.
> 
> This PR is a try on creating a DigitalIOManager that also is able to get turnouts, sensors and lights. I needed to create a connection and wasn't sure on how to do it, so maybe there are better was to do it?
> 
> Any thoughts of this before I continue with AnalogIO and StringIO?
> 
I can see a use for this type of thing on systems where the IO pins could e multiple things, such as XBee and C/MRI.  

On the XBee, there are pins that can be configured as digital inputs, digital outputs, PWM ( analog ) outputs, and  analog inputs.  On C/MRI, I believe the options are a little more limited.

Where this is useful is in preventing duplicate pin assignments.  This is already something we do on XBee, but there it is specifically a per-node check.

This is less useful on other systems where the specific hardware determines the output type.

danielb987 commented:
The use case for me right now is the Multi_IO_Device for LocoNet mentioned in #6741.

danielb987 commented:
I have now added managers for AnalogIO and StringIO. I followed the same concept as for DigitalIO in case someone implements specialized managers for these. For example, #6238 discusses letting MultiMeter extend AnalogIO. Maybe the solution for that is to create the interface Meter that extends AnalogIO and then create a MeterManager.

Another issue is which device type letters AnalogIO, DigitalIO and StringIO should have. I have letters A, D and S for now, but I suppose this needs to change before the PR is merged?

pabender commented:
Sent from my iPad

> On Mar 31, 2019, at 3:55 PM, Daniel <notifications@github.com> wrote:
> I have now added managers for AnalogIO and StringIO. I followed the same concept as for DigitalIO in case someone implements specialized managers for these. For example, #6238 discusses letting MultiMeter extend AnalogIO. Maybe the solution for that is to create the interface Meter that extends AnalogIO and then create a MeterManager.
> 
We already have a MultiMeter object.  I see no benefit to adding another object between MultiMeter and Meter.
> Another issue is which device type letters AnalogIO, DigitalIO and StringIO should have. I have letters A, D and S for now, but I suppose this needs to change before the PR is merged?
> 

You can’t use S.  S is a sensor.

I after reading this, I am wondering if this level of abstraction really helps us.

danielb987 commented:
> We already have a MultiMeter object. I see no benefit to adding another object between MultiMeter and Meter.

MultiMeter has two values, voltage and current. AnalogIO has only one value. In order to let MultiMeter implement AnalogIO, you have to change MultiMeter to only have one meter instead of two. If you don't, you wouldn't know if the value from MultiMeter.getKnownAnalogValue() refers to the voltage or the current.

My point is that it would be better to create a new interface, Meter, that inherits AnalogIO, and then use that as a replacement for MultiMeter.

Pugwash1 commented:
A multimeter sensor has many values. It should supply min, max, value, percent, a string describing the units, what it's measuring, from what device. The new Digitrax units supply something in the order of up to 15 values per device, that's dcs's, bpx88s, bxpa1s all supply stats and measurements, then run dcc++ with a stand alone loconet....
The interface should allow the user to create a dash board from these, I 've been trying to make a panel object, like the fast clock, but not getting anywhere.
Steve G


On March 31, 2019 4:40:29 PM EDT, Daniel <notifications@github.com> wrote:
>> We already have a MultiMeter object. I see no benefit to adding
>another object between MultiMeter and Meter.
>
>MultiMeter has two values, voltage and current. AnalogIO has only one
>value. In order to let MultiMeter implement AnalogIO, you have to
>change MultiMeter to only have one meter instead of two. If you don't,
>you wouldn't know if the value from MultiMeter.getKnownAnalogValue()
>refers to the voltage or the current.
>
>My point is that it would be better to create a new interface, Meter,
>that inherits AnalogIO, and then use that as a replacement for
>MultiMeter.
>
>-- 
>You are receiving this because you are subscribed to this thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/JMRI/JMRI/pull/6757#issuecomment-478380906

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

danielb987 commented:
> A multimeter sensor has many values. It should supply min, max, value, percent, a string describing the units, what it's measuring, from what device. The new Digitrax units supply something in the order of up to 15 values per device, that's dcs's, bpx88s, bxpa1s all supply stats and measurements, then run dcc++ with a stand alone loconet.... The interface should allow the user to create a dash board from these, I 've been trying to make a panel object, like the fast clock, but not getting anywhere. Steve G

The problem with that is that if someone creates something for AnalogIO, for example a Logix Conditional or a control to be used on panels, the MultiMeter cannot use it.

The reason for why I started to work on AnalogIO and StringIO is that I'm working on a successor of Logix, LogixNG, which will have support for both AnalogIO and StringIO. And it would be good if it also supported meters (by having Meter extend AnalogIO).

A useful example is to have a LogixNG monitoring all the meters to see if a booster is deliviering too much current, which could signal that additional boosters are recommended.

Or to monitoring voltage and count the number of short cuts, to see if there are booster districts that has a lot of problems with short cuts.

danielb987 commented:
I have changed the device letters to:
V - AnalogIO (think Variable)
I - dIgitalIO
Z - ZtringIO

I also added Q for LogixNG which I'm working on, and also reordered the letters in alpha order so it's easier to find unused letters when new things are added.

danielb987 commented:
Maybe @Pugwash1 is right, that it would be good if MultiMeter could have many values and that it has support for the user to create a dash board for these values.

My point is that it would be good if all these values _also_ are accessible individually as AnalogIOs to be used for example by LogixNG.

danielb987 commented:
This PR is ready to be merged unless anyone has more comments to it.

danielb987 commented:
I would apreciate if this PR could be merged. Thanks in advance.

danielb987 commented:
Added a space to restart Travis due to:

Failures: 
JsonTimeSocketServiceTest.testOnMessage:67 Time is current expected:<2019-04-05T06:01:02.[64]1+0000> but was:<2019-04-05T06:01:02.[52]1+0000>

The problem is fixed in PR #6773.

danielb987 commented:
AppVeyor failed due to:
WARN  - testLearnWarrant(jmri.jmrit.logix.LearnWarrantTest) : run  1 failed, RetryRule repeats [main]

danielb987 commented:
This PR is ready to be merged.

bobjacobsen commented:
I've been working through this PR in some detail.  I don't think I understand the architecture of it, particularly around the "adapter" idea, but it certainly works: I've been able to run some pretty complicated JMRI instances on this branch without problems. Going ahead with the merge; future conversations might want to move to jmri-developers.
